### PR TITLE
NIFI-2909 and NIFI-1712 Per-Instance ClassLoading

### PR DESCRIPTION
**Pull request by bbende**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:
### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
   in the commit message?
- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [X] Is your initial contribution a single, squashed commit?
### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
### Note:

Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
## Summary of Changes
- Introduced a new annotation for processors, reporting tasks, controller services - @RequiresInstanceClassLoading
- Added new builder method to PropertyDescriptor dynamicallyModifiesClasspath(boolean)
- Created a new InstanceClassLoader that maintains a child-first internal ClassLoader where classpath resources can be set, only used when component has @RequiresInstanceClassLoading
- Added support to ExtensionManager to obtain ClassLoaders based on type + id, before we only had by type
- If the type being requested supports instance class loading (the annotation described above) it will create an InstanceClassLoader which starts as a copy of the NAR ClassLoader for that type
- When properties are set on a component, the framework will identify the properties that are classpath modifiers and add those resources to the InstanceClassLoader's child resources

### CODE REVIEW

**Comment by olegz** [comment]
Reviewing. . .

**Comment by olegz** [comment]
@bbende While i am still reviewing and testing, I think it is worth clarifying what per-instance really means. What I am trying to say is the fact that property values can be changed multiple times per single instance of the component and while it's ok for most properties, changing values for things like additional classpath will most likely result in unpredictable type resolution exceptions.
For example; If a JAR provided as property value is changed after a component already had a chance to run and certain classes were already loaded from the previous JAR while new versions (compatible or not) of the same classes are coming with the new JAR.

Alternatively, we can take a slightly different approach and address it similarly to the way it is addressed in SpringContextProcessor where additional classpath has no relation to the instance of the component and only exists while component is in active state (e.g., CS-enabled, Proc-started, etc.)

To summarize; we either have to document the limitations and side-effects of the per-instance approach or change it to per-activation.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java [hunk 1]
```diff
@@ -90,47 +107,80 @@ public void setAnnotationData(final String data) {
...
         } finally {
             lock.unlock();
         }
     }
 
+    // Keep setProperty/removeProperty private so that all calls go through setProperties
+    private void setProperty(final String name, final String value) {
+        if (null == name || null == value) {
+            throw new IllegalArgumentException();
+        }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: Hmm. . . No message in the exception? Also, should we even check for nulls here given that it's private? meaning we have full control of when/how/where it is invoked?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java [hunk 2]
```diff
@@ -90,47 +107,80 @@ public void setAnnotationData(final String data) {
...
+                final ControllerServiceNode newNode = serviceProvider.getControllerServiceNode(value);
+                if (newNode != null) {
+                    newNode.addReference(this);
+                }
+            }
+
+            try {
+                component.onPropertyModified(descriptor, oldValue, value);
+            } catch (final Exception e) {
+                // nothing really to do here...
```
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: Hmm. . . So if user implements this method in Processor and it throws exception don't we have to at the very least notify the user that it failed?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java [hunk 3]
```diff
@@ -141,48 +191,64 @@ public void setProperty(final String name, final String value) {
...
+    private void processClasspathModifiers(final Set<String> modulePaths) {
+        try {
+            final URL[] urls = ClassLoaderUtils.getURLsForClasspath(modulePaths, null, true);
+
+            final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            if (!(classLoader instanceof InstanceClassLoader)) {
+                // Really shouldn't happen, but if we somehow got here and don't have an InstanceClassLoader then log a warning and move on
+                logger.warn("Unable to modify the classpath for {}, expected InstanceClassLoader, but found {}",
+                        new Object[] { name, classLoader.getClass().getName()});
+                return;
```
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: I am wondering if this should really be a fatal condition. Let's say I have a processor that utilizes this mechanism and the additional classpath is a required property (e.g., like we have in new JMS). This means it can *only* work if such classpath is provided and successfully loaded. Letting it continue with simple warning message wil not accomplish anything, right?

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java [hunk 4]
```diff
@@ -52,23 +86,29 @@ public static ClassLoader getCustomClassLoader(String modulePath, ClassLoader pa
...
-                            File[] files = modulePath.listFiles(filenameFilter);
+                            if (modulePath.isDirectory()) {
+                                File[] files = modulePath.listFiles(filenameFilter);
 
-                            if (files != null) {
-                                for (File jarFile : files) {
-                                    additionalClasspath.add(jarFile.toURI().toURL());
+                                if (files != null) {
+                                    for (File jarFile : files) {
+                                        additionalClasspath.add(jarFile.toURI().toURL());
```
Comment (bbende) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java]: Valid point, the current behavior is the module path can be a file, or a directory with files, but it doesn't recurse further into sub-directories. This was existing functionality that was written for the scripting processors that I was just re-using: https://github.com/apache/nifi/blob/master/nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java [hunk 5]
```diff
@@ -24,23 +24,57 @@
...
+                        .filter(StringUtils::isNotBlank)
+                        .map(String::trim)
+                        .collect(Collectors.toSet()));
+            }
+        }
+
+        return getURLsForClasspathHelper(modules, filenameFilter, suppressExceptions);
+    }
+
+    protected static URL[] getURLsForClasspathHelper(Set<String> modulePaths, FilenameFilter filenameFilter, boolean suppressExceptions) throws MalformedURLException {
```
Comment (bbende) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java]: Sure that would make sense.

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java [hunk 6]
```diff
@@ -24,23 +24,57 @@
...
+            for (String modulePath : modulePaths) {
+                modules.addAll(Arrays.stream(modulePath.split(","))
+                        .filter(StringUtils::isNotBlank)
+                        .map(String::trim)
+                        .collect(Collectors.toSet()));
+            }
+        }
+
+        return getURLsForClasspathHelper(modules, filenameFilter, suppressExceptions);
+    }
```
Comment (bbende) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java]: Sure I'm open to cleaning up the code here.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java (hunk 1 above)
Comment (bbende) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: That does seem weird, I was preserving existing behavior here:
https://github.com/apache/nifi/blob/d838f61291d2582592754a37314911b701c6891b/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java#L95

**Comment by bbende** [comment]
@olegz thanks for reviewing, I posted a couple of replies to the inline comments.

I also wanted to mention that after submitting this PR, I realized that a slight modification to the approach might open this feature up to a wider variety of use cases...

In the current PR, when the component is annotated with @RequiresInstanceClassLoading it creates an InstanceClassLoader which copies all of the URL resources from the NAR ClassLoader. This was necessary to solve the HBase + Phoenix problem, but it many other cases it may be acceptable to just make a new InstanceClassLoader with a parent of the NAR ClassLoader.

So I started working on a change where every single processor, cs, reporting task will get an InstanceClassLoader and by default it will be empty and have a parent of the NAR ClassLoader. Only components with the @RequiresInstanceClassLoading will force a full copy of the resources. 

This should lets the HBaseControllerService use @RequiresInstanceClassLoading, and other components can still have properties that modify the classpath without making full copies.

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java (hunk 4 above)
Comment (olegz) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java]: That's fine and is generally expected for classpath entries. However, I would consider NOT adding it to the ```additionalClasspath``` while logging a warning stating that recursive classpath hierarchies are not supported

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java (hunk 3 above)
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: Agree with the first scenario.
As for the ```ClassLoaderUtils.getURLsForClasspath(...)```, I've read what you describe and I personally lean to the side of fail fast. Don't get me wrong; what you describe is perfectly valid, but I approach it very differently where the _fatal_ condition I am referring to will most definitely be a flow developer's error which will be caught during the development of the flow and therefore is not expected to happen in prod, so delegating to the component developer to do custom validation in each and every new/existing component seems like on overhead.

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java (hunk 4 above)
Comment (bbende) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/classloader/ClassLoaderUtils.java]: I see what you mean, I'll add that check to see if it is a directory or file and handle accordingly

**Comment by bbende** [comment]
@olegz I pushed up a second commit that with the change that I described above, and also some of the changes from your feedback. Unfortunately I had already rebased my branch before you had started reviewing, so i ended up having to force push, but I left it as two commits where the first one has the stuff you already looked at, and the second commit has changes after that.

I think the remaining discussion point is mostly around how restrictive we get regarding what happens when one or more resources can't be found. I understand the point about what happens if something that is necessary can't be loaded, but I also want to give flexibility to the developers using this feature. There may be cases where someone wants to provide a couple of directories that are common locations for a given library, and if found then load whatever is found, but if not then maybe they don't care. One option might be to provide a validator that uses ClassLoader.getURLsforClasspath to validate everything resolves, and then document that this is available, but not enforce that it is always used.

**Comment by olegz** [comment]
@bbende thanks, will start looking

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java (hunk 1 above)
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: I see, but I still think we have to fix it

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java (hunk 2 above)
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: I see, but I still think we have to fix it (i know, it's the same comment as above ;) )

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java [hunk 7]
```diff
@@ -141,48 +181,74 @@ public void setProperty(final String name, final String value) {
...
 
-                    try {
-                        component.onPropertyModified(descriptor, value, null);
-                    } catch (final Exception e) {
-                        // nothing really to do here...
-                    }
+            try {
+                component.onPropertyModified(descriptor, value, null);
+            } catch (final Exception e) {
+                // nothing really to do here...
```
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java]: Same as above, just doesn't look right

**Comment by olegz** [comment]
@bbende just left a few minor comments which probably all fall in the category of stylistic, so it's up to you to decide if you want to change anything. 
Now I need to spend some time and play with it to see exactly how it works. Will do it over the weekend.

File: nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java [hunk 8]
```diff
@@ -67,6 +67,14 @@
             .defaultValue("1")
             .build();
 
+    PropertyDescriptor PHOENIX_CLIENT_JAR_LOCATION = new PropertyDescriptor.Builder()
+            .name("Phoenix Client JAR Location")
+            .description("The full path to the Phoenix client JAR. Required if Phoenix is installed on top of HBase.")
+            .addValidator(StandardValidators.FILE_EXISTS_VALIDATOR)
+            .expressionLanguageSupported(true)
+            .dynamicallyModifiesClasspath(true)
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: There are other properties (such as Module Directory in ExecuteScript and InvokeScriptedProcessor, and one in JoltTransformJSON) that add JARs to the classpath, should they be updated as part of this PR to set "dynamicallyModifiesClasspath" to true?

Comment (bbende) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: They could be... my thinking was to initially get this capability in place and then maybe the people who worked on those components could determine if they should be refactored to use this capability, or if there is some reason why they should be left as is.

Comment (olegz) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: @bbende but we won't be able to do most of it for existing components until we are working on the new major release, correct? What I mean is that those would be breaking changes.

Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: What do you mean by "breaking"? If it changes the property such that the processor becomes invalid, I think we are allowed to do that (the user just has to remove the old property), as long as the default behavior is the same (which it will be in this case since the new feature defaults to false). In any case I'm good with writing a separate Jira to consider the other components that do classloading stuff

Comment (olegz) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: Well, that is what I am trying to figure out. But regardless, _breaking_ simply means that without doing something extra it will no longer work. If this is the case then it's breaking, correct. I do understand that certain cases may be allowed for greater good and of course we would need to document them, but I just wanted to be clear on the implications.

Comment (bbende) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: I don't see why it would be considered breaking... in most of the cases nothing is changing for the end user. Lets take DBCP Connection Pool, the same property exists before and after, its just the class loading code being moved from in the component to handled by the framework.

Comment (bbende) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase-client-service-api/src/main/java/org/apache/nifi/hbase/HBaseClientService.java]: and to Matt's point, even if it did make the property become invalid on upgrade, that is considered acceptable on a minor release.

**Comment by bbende** [comment]
@olegz just pushed up a commit with some changes from the last round of feedback (renaming jarFile, fixing empty catch block, fixing empty IllegalArgumentException).

Have you had a chance to test this out? 

Also let me know if/when you want me to squash the commits

**Comment by olegz** [comment]
@bbende I am actually in the process of testing it now, but giving that this is class loading, give till tomorrow and if all is good I'll merge. You can squash now, I am good with the changes you've made.
Also, probably lost in the maze of comments, but i don't think we ever closed the loop on per-instance vs, per-lifecycle point i've made earlier. What's your take on that?

**Comment by bbende** [comment]
@olegz cool thanks! regarding the per-instance vs. per-lifecycle, in this first pass we are basically loading the resources at the time the properties are applied in order to be available during validation, I think in the future we can determine if we need something additional that is more tied to the component lifecycle.

**Comment by olegz** [comment]
@bbende I now tend to agree. Per-lifecycle class loading is an edge case and we only use it right now in SpringContextProcessor where it is required due to the fact that Spring itself is a developer framework and it's class path heavily depends on volatile artifacts (e.g., configuration files, application code etc.). So while most JARs will remain the same, the overall classpath may change after re-starts. For the cases where one deals with JMS ConectionFactories and JDBS Drivers that most likely never be the case.

**Comment by b-long** [comment]
@olegz @bbende , et al.  Any insight on this -> https://stackoverflow.com/questions/42938158/nifi-springcontextprocessor-using-configuration-instead-of-xml#comment84380723_42946116 ?
